tests/generation/test_utils.py

```diff
@@ -2730,6 +2730,20 @@ def test_max_length_warning_if_different(self):
                 **model_kwargs,
             )
 
+    def test_max_length_if_input_embeds(self):
+        # PT-only test: TF doesn't have StoppingCriteria
+        article = "Hey, are you conscious?"
+        model = AutoModelForCausalLM.from_pretrained("hf-internal-testing/tiny-random-gpt2").to(torch_device)
+        tokenizer = AutoTokenizer.from_pretrained("hf-internal-testing/tiny-random-gpt2")
+        input_ids = tokenizer(article, return_tensors="pt").input_ids.to(torch_device)
+        inputs_embeds = model.get_input_embeddings()(input_ids)
+
+        max_length = 20
+        input_len = input_ids.shape[-1]
+        out_gen = model.generate(input_ids=input_ids, max_length=max_length)
+        out_gen_embeds = model.generate(inputs_embeds=inputs_embeds, max_length=max_length)
+        self.assertEqual(out_gen.shape[-1], input_len + out_gen_embeds.shape[-1])
```

Comment:
For my own understanding - why is the returned generation when passing in input_ids, a concatenation of the input and newly generated tokens, but for embeds we only return the new embeddings?

Reply:
The addition of input_length here is needed because the output of generation with `inputs_embeds` return only newly generated text, while the `input_ids` return the whole text, including prompt. So, we are just making sure the lengths of both are equal

Reply:
Right, but why is the behaviour different for embeddings and input_ids?

Reply:
If I understand the question correctly, the lengths here differ because we return the whole text (prompt+new) when user passes `ids`. But we cannot recover prompt text from `input_embeds`, so we just return the newly generated part

Reply:
As @zucchini-nlp wrote. 

There is no mismatch if the user passes `input_ids` and `inputs_embeds`, as `generate` continues populating `input_ids`. But passing both kinda defeats the point of feeding `inputs_embeds`, which is used mostly for experimental purposes, and thus the shape difference when only `inputs_embeds` is set. Although we technically can recover `input_ids` from `inputs_embeds` in most cases (reverse lookup search) to make the shapes consistent, it's probably not a good use of our engineering time :D

Reply:
@zucchini-nlp @gante Thanks for the explanation!